File: RickAndMortyAPP/app/src/main/java/com/example/lpiem/rickandmortyapp/RickAndMortyRetrofitSingleton.java
package com.example.lpiem.rickandmortyapp;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RickAndMortyRetrofitSingleton {

    static final String BASE_URL = "https://rickandmortyapi.com/";

    private static RickAndMortyAPI rickAndMortyAPIInstance;

    public static RickAndMortyAPI getInstance(){
        if(rickAndMortyAPIInstance == null)
            synchronized (RickAndMortyAPI.class){
            createApiBuilder();
            }
            return rickAndMortyAPIInstance;
    }

    private static void createApiBuilder(){
        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
        rickAndMortyAPIInstance = retrofit.create(RickAndMortyAPI.class);
    }
}

File: RickAndMortyAPP/app/src/main/java/com/example/lpiem/rickandmortyapp/RickAndMortyRetrofitSingleton.kt
package com.example.lpiem.rickandmortyapp

import com.google.gson.GsonBuilder

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RickAndMortyRetrofitSingleton {

    private const val BASE_URL = "https://rickandmortyapi.com/"

    private var rickAndMortyAPIInstance: RickAndMortyAPI? = null

    val instance: RickAndMortyAPI?
        get() {
            if (rickAndMortyAPIInstance == null)
                synchronized(RickAndMortyAPI::class.java) {
                    createApiBuilder()
                }
            return rickAndMortyAPIInstance
        }

    private fun createApiBuilder() {
        val gsonInstance = GsonBuilder()
                .setLenient()
                .create()

        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gsonInstance))
                .build()
        rickAndMortyAPIInstance = retrofit.create(RickAndMortyAPI::class.java)
    }
}
